### PR TITLE
Prerequisites for adding Thinkerbell as an adapter.

### PR DESCRIPTION
**Pull request by mcav**

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -113,6 +113,10 @@ pub enum InternalError {
     /// Attempting to register a channel with an adapter that doesn't match that of its service.
     ConflictingAdapter(Id<AdapterId>, Id<AdapterId>),
 
+    /// Open question: Individual adapters will have errors of many adapter-specific types.
```
Comment (Yoric) [inline src/api.rs:131]: Maybe by making it `JSON`? Not sure.

File: src/services.rs [hunk 2]
```diff
@@ -158,6 +158,9 @@ pub enum ChannelKind {
     Thermostat,
     ActualTemperature,
 
+    AddRule,
```
Comment (Yoric) [inline src/services.rs]: I'd prefer `AddThinkerbellRule`/`RemoveThinkerbellRule`. Also, doc would be useful.

File: src/values.rs [hunk 3]
```diff
@@ -431,6 +433,8 @@ pub enum Value {
 
     // FIXME: Add more as we identify needs
 
+    ThinkerbellRule(String, String),
```
Comment (Yoric) [inline src/values.rs]: Strings are ambiguous. I'd prefer if it were a `struct`. I assume this is `(name, source)`?

**Comment by mcav** [comment]
@Yoric Updated with the following changes:
- AddRule -> AddThinkerbellRule
- RemoveRule was deleted, turns out it wasn't used
- Changed ThinkerbellRule enum to use a struct instead of `String, String`

I have not changed the GenericError object, as it seems like that could be reasonably addressed in a followup.

**Comment by Yoric** [comment]
> RemoveRule was deleted, turns out it wasn't used

Weird. How do you handle rule removal?

> I have not changed the GenericError object, as it seems like that could be reasonably addressed in a followup.

Makes sense.

File: src/values.rs [hunk 4]
```diff
@@ -632,6 +635,32 @@ impl ToJSON for Color {
...
+impl Parser<ThinkerbellRuleSource> for ThinkerbellRuleSource {
+    fn parse(path: Path, source: &mut JSON) -> Result<Self, ParseError> {
+        let name = try!(path.push("name", |path| String::take(path, source, "name")));
+        let script_source = try!(path.push("source", |path| String::take(path, source, "source")));
+        Ok(ThinkerbellRuleSource { name: name, source: script_source })
+    }
+}
+impl ToJSON for ThinkerbellRuleSource {
+    fn to_json(&self) -> JSON {
+        let mut source = vec![
```
Comment (Yoric) [inline src/values.rs]: Actually, I have added stuff that should now let you simply write

``` rust
fn to_json(&self) -> JSON {
  vec![...].to_json()
}
```

Can be done later, though, if you prefer.

**Comment by Yoric** [comment]
Looks good to me, with the remarks above.

**Comment by mcav** [comment]
> How do you handle rule removal?

The 'remove' setter is currently of type `ChannelKind::Ready`. It doesn't need any associated data, because every setter maps to exactly one Rule. We could add another ChannelKind, I guess, but the other ChannelKind values are quite generic so I wasn't sure which route we wanted to take.

**Comment by Yoric** [comment]
I'd go for something more specific. When the client wishes to remove a rule, they'll look for all the getters of `ChannelKind::RemoveThinkerbellRule`.

Of course, we still need a good story for determining which rule they are removing. I don't know how we can handle that properly yet.

**Pull request merged**